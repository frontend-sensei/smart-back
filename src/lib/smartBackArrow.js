export class SmartBackArrow {
  constructor(instance) {
    this.instance = instance;
    this.isActive = false;
  }

  setActive() {
    this.instance._movement.needSkipEvent = true;

    this.arrowStateHandler("add");
    this.updateTransition(
      `transform .${this.instance.TRANSITION_DURATION}s ease-in-out`
    );
    this.instance._arrow.updatePosition();
    setTimeout(() => {
      this.updateTransition("unset");
      this.instance._movement.needSkipEvent = false;
      this.isActive = true;
    }, this.instance.TRANSITION_DURATION);
  }
  setInactive() {
    this.instance._movement.needSkipEvent = true;

    this.arrowStateHandler("remove");
    this.resetValues();
    setTimeout(() => {
      this.instance._movement.needSkipEvent = false;
      this.isActive = false;
    }, this.instance.TRANSITION_DURATION);
  }
  arrowStateHandler(actionName) {
    this.arrowClassHandler({
      actionName,
      className: this.instance.ENABLE_ARROW_MIRRORING
        ? `smart-back__arrow--active-from-${this.instance._movement.currentSide}`
        : "smart-back__arrow--active",
    });
  }
  arrowClassHandler({ actionName, className }) {
    this.classHandler({
      id: this.instance.arrowId,
      actionName,
      className,
    });
  }
  classHandler({ id, actionName = "", className }) {
    const element = document.getElementById(id);
    if (!element) {
      throw new Error(`Element with id: "${id}" not found`);
    }

    const classListHandler = element.classList[actionName];
    if (!classListHandler) {
      throw new Error(`ClassList handler with name: "${actionName}" not found`);
    }

    document.getElementById(id).classList[actionName](className);
  }
  updatePosition() {
    this.setStyle(
      "transform",
      `translate(${this.instance._movement.translateX}px, ${this.instance._movement.translateY}px)`
    );
  }
  setStyle(style, value) {
    this.updateNodeStyle(style, value);
  }
  updateNodeStyle(style, value) {
    this.instance._nodes.arrowNode.style[style] = value;
  }
  updateTransition(value) {
    this.updateCSSVar("--sb-arrow-transition", value);
  }
  updateCSSVar(name, value) {
    document.documentElement.style.setProperty(name, value);
  }
  resetValues() {
    this.updateTransition("transform .1s ease-in-out");
    this.setStyle("transform", "translate(0,0)");
    setTimeout(() => {
      this.instance._movement.resetValues();
      this.updateTransition("unset");
    }, this.instance.TRANSITION_DURATION);
  }
}
