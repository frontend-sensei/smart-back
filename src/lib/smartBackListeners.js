export class SmartBackListeners {
  constructor(instance) {
    this.instance = instance;
  }

  touchstart(event) {
    const currentSide = event.target.dataset.side;
    if (!currentSide) {
      throw new Error("Attribute data-side not found");
    }

    this.instance._listenersHandling.registerAdditionalListeners();
    this.instance._movement.currentSide = currentSide;

    this.instance._movement.prevClientX = event.touches[0].clientX;
    this.instance._movement.prevClientY = event.touches[0].clientY;

    this.instance._arrow.setStyle("top", `${event.touches[0].clientY - 55}px`);
  }
  touchmove(event) {
    event.preventDefault();
    if (event.touches.length > 1) {
      this.instance._listenersHandling.removeAdditionalListeners();
      return;
    }

    if (this.instance._movement.needSkipEvent) {
      return;
    }

    this.instance._arrow.setStyle("left", "unset");
    this.instance._arrow.setStyle("right", "unset");
    this.instance._arrow.setStyle(this.instance._movement.currentSide, "-4px");

    this.instance._movement.getMovementDirections(event);
    this.instance._movement.movementAlongXHandler();
  }
  touchend() {
    if (this.instance._arrow.isActive) {
      this.instance.CALLBACK();
    }
    this.resetToInitialState();
  }
  touchcancel() {
    this.resetToInitialState();
  }
  resetToInitialState() {
    this.instance._listenersHandling.removeAdditionalListeners();
    this.instance._arrow.setInactive();
  }
}
