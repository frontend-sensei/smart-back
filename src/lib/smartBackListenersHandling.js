export class SmartBackListenersHandling {
  constructor(instance) {
    this.instance = instance;
    this.boundListeners = {
      touchstart: instance._listeners.touchstart.bind(instance._listeners),
    };
    this.boundAdditionalListeners = {
      touchmove: instance._listeners.touchmove.bind(instance._listeners),
      touchend: instance._listeners.touchend.bind(instance._listeners),
      touchcancel: instance._listeners.touchcancel.bind(instance._listeners),
    };
    this.registerListeners();
  }

  registerListeners() {
    this.listenersHandler("addEventListener");
  }
  removeListeners() {
    this.listenersHandler("removeEventListener");
  }
  listenersHandler(listenerAction) {
    this.instance._nodes.sideNodes.left[listenerAction](
      "touchstart",
      this.boundListeners.touchstart,
      false
    );
    this.instance._nodes.sideNodes.right[listenerAction](
      "touchstart",
      this.boundListeners.touchstart,
      false
    );
  }

  registerAdditionalListeners() {
    this.additionalListenersHandler(
      "addEventListener",
      this.instance._nodes.sideNodes.left
    );
    this.additionalListenersHandler(
      "addEventListener",
      this.instance._nodes.sideNodes.right
    );
  }
  removeAdditionalListeners() {
    this.additionalListenersHandler(
      "removeEventListener",
      this.instance._nodes.sideNodes.left
    );
    this.additionalListenersHandler(
      "removeEventListener",
      this.instance._nodes.sideNodes.right
    );
  }
  additionalListenersHandler(action, sideNode) {
    if (!sideNode) {
      throw new Error("Side node not found!");
    }
    sideNode[action](
      "touchmove",
      this.boundAdditionalListeners.touchmove,
      false
    );
    sideNode[action]("touchend", this.boundAdditionalListeners.touchend, false);
    sideNode[action](
      "touchcancel",
      this.boundAdditionalListeners.touchcancel,
      false
    );
  }
}
