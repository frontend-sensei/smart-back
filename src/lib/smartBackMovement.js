export class SmartBackMovement {
  constructor(instance) {
    this.instance = instance;
    this.needSkipEvent = false;
    this.currentSide = "";
    this.currentDirections = {
      horizontal: "",
      vertical: "",
    };
    this.prevClientX = 0;
    this.prevClientY = 0;
    this.translateX = 0;
    this.translateY = 0;
    this.variationAlongX = 0.4;
    this.variationAlongY = 0.4;
    this.movedToCancelAlongX = 0;
    this.isNeedBreak = false;
    this.variationLimit = 0.045;
    this.subtrahendValueAlongY = 0.002;
    this.subtrahendValueAlongX = 0.002;
  }

  getMovementDirections(event) {
    const { clientX, clientY } = event.touches[0];

    if (this.prevClientX < clientX) {
      this.currentDirections.horizontal = "right";
    } else if (this.prevClientX > clientX) {
      this.currentDirections.horizontal = "left";
    } else if (this.prevClientX === clientX) {
      this.currentDirections.horizontal = "";
    }

    if (this.prevClientY < clientY) {
      this.currentDirections.vertical = "bottom";
    } else if (this.prevClientY > clientY) {
      this.currentDirections.vertical = "top";
    } else if (this.prevClientY === clientY) {
      this.currentDirections.vertical = "";
    }

    this.prevClientX = clientX;
    this.prevClientY = clientY;
  }

  movementAlongXHandler() {
    this.isNeedBreak = false;

    if (this.currentSide === "right") {
      if (this.currentDirections.horizontal === "left") {
        this.movementTowardsActivation(
          this.translateX - this.variationAlongX,
          this.translateX > -this.instance.ARROW_TRIGGERING_OFFSET,
          -this.instance.STATIC_ACTIVE_TRANSLATE_X
        );
      }
      if (this.currentDirections.horizontal === "right") {
        this.movementTowardsDeactivation(
          this.translateX + this.variationAlongX
        );
      }
    }

    if (this.currentSide === "left") {
      if (this.currentDirections.horizontal === "right") {
        this.movementTowardsActivation(
          this.translateX + this.variationAlongX,
          this.translateX < this.instance.ARROW_TRIGGERING_OFFSET,
          this.instance.STATIC_ACTIVE_TRANSLATE_X
        );
      }
      if (this.currentDirections.horizontal === "left") {
        this.movementTowardsDeactivation(
          this.translateX - this.variationAlongX
        );
      }
    }

    if (this.instance._arrow.isActive && !this.isNeedBreak) {
      this.movementAlongYHandler();
    }
  }
  movementAlongYHandler() {
    if (this.currentDirections.vertical === "top") {
      this.translateY -= this.variationAlongY;
    } else if (this.currentDirections.vertical === "bottom") {
      this.translateY += this.variationAlongY;
    }
    this.instance._arrow.updatePosition();

    if (this.variationAlongY < this.variationLimit) {
      return;
    }
    this.variationAlongY -= this.subtrahendValueAlongY;
  }

  movementTowardsActivation(
    translateX,
    isTranslateXReachedOffset,
    activeTranslateX
  ) {
    this.movedToCancelAlongX = 0;
    this.translateX = translateX;

    if (isTranslateXReachedOffset) {
      return;
    }

    if (!this.instance._arrow.isActive) {
      this.translateX += activeTranslateX;
      this.instance._arrow.setActive();
      return;
    }

    this.instance._arrow.updatePosition();

    if (this.variationAlongX < this.variationLimit) {
      return;
    }
    this.variationAlongX -= this.subtrahendValueAlongX;
  }
  movementTowardsDeactivation(translateX) {
    if (!this.instance._arrow.isActive) {
      return this.needBreak();
    }

    if (this.movedToCancelAlongX === this.instance.ARROW_TRIGGERING_OFFSET) {
      this.instance._arrow.setInactive();
      return this.needBreak();
    }

    this.movedToCancelAlongX += 1;

    if (this.movedToCancelAlongX < this.instance.ARROW_TRIGGERING_OFFSET) {
      this.translateX = translateX;
      this.instance._arrow.updatePosition();
    }
  }

  needBreak() {
    this.isNeedBreak = true;
  }
  resetValues() {
    this.variationAlongX = 0.4;
    this.variationAlongY = 0.4;
    this.translateX = 0;
    this.translateY = 0;
    this.movedToCancelAlongX = 0;
  }
}
