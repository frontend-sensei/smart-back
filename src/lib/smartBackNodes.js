export class SmartBackNodes {
  constructor(instance) {
    this.instance = instance;
    this.sideNodes = {
      left: null,
      right: null,
    };
    this.arrowNode = null;
    this.getNodes();
  }

  getNodes() {
    this.sideNodes = this.getSideNodes();
    this.arrowNode = this.getArrowNode();
  }
  getSideNodes() {
    const left = document.getElementById(this.instance.leftSideId);
    const right = document.getElementById(this.instance.rightSideId);
    if (!left) {
      throw new Error("Left side node not found!");
    }
    if (!right) {
      throw new Error("Right side node not found!");
    }

    return {
      left,
      right,
    };
  }
  getArrowNode() {
    const arrowNode = document.getElementById(this.instance.arrowId);
    if (!arrowNode) {
      throw new Error("Arrow node not found!");
    }
    return arrowNode;
  }
}
