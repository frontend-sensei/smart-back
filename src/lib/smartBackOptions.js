export class SmartBackOptions {
  constructor(options) {
    this.options = options;
  }

  validate(propertyName, reserveValue) {
    if (this.options.hasOwnProperty(propertyName)) {
      return this.options[propertyName];
    }
    return reserveValue;
  }
}
