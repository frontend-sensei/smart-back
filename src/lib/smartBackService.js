export class SmartBackService {
  constructor(
    options = {},
    vendors = {
      _optionsClass: null,
      _uiClass: null,
      _nodesClass: null,
      _listenersClass: null,
      _listenersHandlingClass: null,
      _arrowClass: null,
      _movementClass: null,
    }
  ) {
    this._options = new vendors._optionsClass(options);

    this.classes = {
      wrapper: "smart-back",
      side: "smart-back__side",
      sideLeft: "smart-back__side--left",
      sideRight: "smart-back__side--right",
    };
    this.leftSideId = "smart_back_side_left";
    this.rightSideId = "smart_back_side_right";
    this.arrowId = "smart_back_arrow";

    this.TRANSITION_DURATION = this._options.validate(
      "transitionDuration",
      140
    );
    this.ARROW_TRIGGERING_OFFSET = this._options.validate(
      "arrowTriggeringOffset",
      2
    );
    this.STATIC_ACTIVE_TRANSLATE_X = this._options.validate(
      "staticActiveTranslateX",
      40
    );
    this.ENABLE_ARROW_MIRRORING = this._options.validate(
      "enableArrowMirroring",
      true
    );
    this.VIBRATION = this._options.validate("vibration", 10);
    this.CALLBACK = this._options.validate(
      "callback",
      history.back.bind(window.history)
    );

    this._ui = vendors._uiClass;
    this._nodes = vendors._nodesClass;
    this._listeners = vendors._listenersClass;
    this._listenersHandling = vendors._listenersHandlingClass;
    this._arrow = vendors._arrowClass;
    this._movement = vendors._movementClass;

    this.render();
  }

  render() {
    this._ui = new this._ui(this);
    this._nodes = new this._nodes(this);
    this._listeners = new this._listeners(this);
    this._listenersHandling = new this._listenersHandling(this);
    this._arrow = new this._arrow(this);
    this._movement = new this._movement(this);
  }

  destroy() {
    this._ui.remove();
    this._listenersHandling.removeListeners();
    this._listenersHandling.removeAdditionalListeners();
  }
}
