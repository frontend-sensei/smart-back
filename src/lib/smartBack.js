import { SmartBackService } from "./smartBackService";

import { SmartBackOptions } from "./smartBackOptions";
import { SmartBackUI } from "./smartBackUI";
import { SmartBackNodes } from "./smartBackNodes";
import { SmartBackListeners } from "./smartBackListeners";
import { SmartBackListenersHandling } from "./smartBackListenersHandling";
import { SmartBackArrow } from "./smartBackArrow";
import { SmartBackMovement } from "./smartBackMovement";

const vendors = {
  _optionsClass: SmartBackOptions,
  _uiClass: SmartBackUI,
  _nodesClass: SmartBackNodes,
  _listenersClass: SmartBackListeners,
  _listenersHandlingClass: SmartBackListenersHandling,
  _arrowClass: SmartBackArrow,
  _movementClass: SmartBackMovement,
};

export class SmartBack {
  constructor(options) {
    return new SmartBackService(options, vendors);
  }
}
