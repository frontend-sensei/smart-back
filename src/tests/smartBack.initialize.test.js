// @ts-nocheck
import { SmartBack } from "../lib/smartBack";

describe("SmartBack", () => {
  it("should renders without options", () => {
    const smartBack = new SmartBack();
    expect(smartBack).toMatchSnapshot();
  });

  it("should renders with options", () => {
    const options = {
      transitionDuration: "150",
      arrowTriggeringOffset: "10",
      staticActiveTranslateX: "50",
      enableArrowMirroring: false,
      callback: () => console.log("Done!"),
    };
    const smartBack = new SmartBack(options);
    expect(smartBack).toMatchSnapshot();
  });
});
