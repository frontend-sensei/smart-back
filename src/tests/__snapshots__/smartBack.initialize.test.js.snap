// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartBack should renders with options 1`] = `
SmartBackService {
  "ARROW_TRIGGERING_OFFSET": "10",
  "CALLBACK": [Function],
  "ENABLE_ARROW_MIRRORING": false,
  "STATIC_ACTIVE_TRANSLATE_X": "50",
  "TRANSITION_DURATION": "150",
  "VIBRATION": 20,
  "_arrow": SmartBackArrow {
    "instance": [Circular],
    "isActive": false,
  },
  "_listeners": SmartBackListeners {
    "instance": [Circular],
  },
  "_listenersHandling": SmartBackListenersHandling {
    "boundAdditionalListeners": Object {
      "touchcancel": [Function],
      "touchend": [Function],
      "touchmove": [Function],
    },
    "boundListeners": Object {
      "touchstart": [Function],
    },
    "instance": [Circular],
  },
  "_movement": SmartBackMovement {
    "currentDirections": Object {
      "horizontal": "",
      "vertical": "",
    },
    "currentSide": "",
    "instance": [Circular],
    "isNeedBreak": false,
    "movedToCancelAlongX": 0,
    "needSkipEvent": false,
    "prevClientX": 0,
    "prevClientY": 0,
    "subtrahendValueAlongX": 0.002,
    "subtrahendValueAlongY": 0.002,
    "translateX": 0,
    "translateY": 0,
    "variationAlongX": 0.4,
    "variationAlongY": 0.4,
    "variationLimit": 0.045,
  },
  "_nodes": SmartBackNodes {
    "arrowNode": <div
      class="smart-back__arrow"
      id="smart_back_arrow"
    >
      
        
      <div
        class="smart-back__arrow-part smart-back__arrow-part-1"
      />
      
        
      <div
        class="smart-back__arrow-part smart-back__arrow-part-2"
      />
      
      
    </div>,
    "instance": [Circular],
    "sideNodes": Object {
      "left": <div
        class="smart-back__side smart-back__side--left"
        data-side="left"
        id="smart_back_side_left"
      />,
      "right": <div
        class="smart-back__side smart-back__side--right"
        data-side="right"
        id="smart_back_side_right"
      />,
    },
  },
  "_options": SmartBackOptions {
    "options": Object {
      "arrowTriggeringOffset": "10",
      "callback": [Function],
      "enableArrowMirroring": false,
      "staticActiveTranslateX": "50",
      "transitionDuration": "150",
      "vibration": 20,
    },
  },
  "_ui": SmartBackUI {
    "instance": [Circular],
    "uiNode": <div
      class="smart-back"
    >
      
      
      <div
        class="smart-back__side smart-back__side--left"
        data-side="left"
        id="smart_back_side_left"
      />
      
      
      <div
        class="smart-back__side smart-back__side--right"
        data-side="right"
        id="smart_back_side_right"
      />
      
      
      <div
        class="smart-back__arrow"
        id="smart_back_arrow"
      >
        
        
        <div
          class="smart-back__arrow-part smart-back__arrow-part-1"
        />
        
        
        <div
          class="smart-back__arrow-part smart-back__arrow-part-2"
        />
        
      
      </div>
      
    
    </div>,
  },
  "arrowId": "smart_back_arrow",
  "classes": Object {
    "side": "smart-back__side",
    "sideLeft": "smart-back__side--left",
    "sideRight": "smart-back__side--right",
    "wrapper": "smart-back",
  },
  "leftSideId": "smart_back_side_left",
  "rightSideId": "smart_back_side_right",
}
`;

exports[`SmartBack should renders without options 1`] = `
SmartBackService {
  "ARROW_TRIGGERING_OFFSET": 2,
  "CALLBACK": [Function],
  "ENABLE_ARROW_MIRRORING": true,
  "STATIC_ACTIVE_TRANSLATE_X": 40,
  "TRANSITION_DURATION": 140,
  "VIBRATION": 10,
  "_arrow": SmartBackArrow {
    "instance": [Circular],
    "isActive": false,
  },
  "_listeners": SmartBackListeners {
    "instance": [Circular],
  },
  "_listenersHandling": SmartBackListenersHandling {
    "boundAdditionalListeners": Object {
      "touchcancel": [Function],
      "touchend": [Function],
      "touchmove": [Function],
    },
    "boundListeners": Object {
      "touchstart": [Function],
    },
    "instance": [Circular],
  },
  "_movement": SmartBackMovement {
    "currentDirections": Object {
      "horizontal": "",
      "vertical": "",
    },
    "currentSide": "",
    "instance": [Circular],
    "isNeedBreak": false,
    "movedToCancelAlongX": 0,
    "needSkipEvent": false,
    "prevClientX": 0,
    "prevClientY": 0,
    "subtrahendValueAlongX": 0.002,
    "subtrahendValueAlongY": 0.002,
    "translateX": 0,
    "translateY": 0,
    "variationAlongX": 0.4,
    "variationAlongY": 0.4,
    "variationLimit": 0.045,
  },
  "_nodes": SmartBackNodes {
    "arrowNode": <div
      class="smart-back__arrow"
      id="smart_back_arrow"
    >
      
        
      <div
        class="smart-back__arrow-part smart-back__arrow-part-1"
      />
      
        
      <div
        class="smart-back__arrow-part smart-back__arrow-part-2"
      />
      
      
    </div>,
    "instance": [Circular],
    "sideNodes": Object {
      "left": <div
        class="smart-back__side smart-back__side--left"
        data-side="left"
        id="smart_back_side_left"
      />,
      "right": <div
        class="smart-back__side smart-back__side--right"
        data-side="right"
        id="smart_back_side_right"
      />,
    },
  },
  "_options": SmartBackOptions {
    "options": Object {},
  },
  "_ui": SmartBackUI {
    "instance": [Circular],
    "uiNode": <div
      class="smart-back"
    >
      
      
      <div
        class="smart-back__side smart-back__side--left"
        data-side="left"
        id="smart_back_side_left"
      />
      
      
      <div
        class="smart-back__side smart-back__side--right"
        data-side="right"
        id="smart_back_side_right"
      />
      
      
      <div
        class="smart-back__arrow"
        id="smart_back_arrow"
      >
        
        
        <div
          class="smart-back__arrow-part smart-back__arrow-part-1"
        />
        
        
        <div
          class="smart-back__arrow-part smart-back__arrow-part-2"
        />
        
      
      </div>
      
    
    </div>,
  },
  "arrowId": "smart_back_arrow",
  "classes": Object {
    "side": "smart-back__side",
    "sideLeft": "smart-back__side--left",
    "sideRight": "smart-back__side--right",
    "wrapper": "smart-back",
  },
  "leftSideId": "smart_back_side_left",
  "rightSideId": "smart_back_side_right",
}
`;
